### PR TITLE
Improve FindPeaksMD

### LINKED ISSUE #19914: Improve FindPeaksMD
`FindPeaksMD` is showing quite bad behaviour in terms of peak finding performance. It finds a lot of peaks which don't exist and does not find quite obvious peaks.
The image below demonstrates this.

full_bad

We need to investigate what the reason for this is and what we can do about this.

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Fixes #19914

~~Currently being tested by Pascal and Fabio~~
The improvements have been evaluated by Pascal and it does work better for him in some situations. Thomas has also reported a similar behaviour. Hence we can go ahead with the changes.

Please see the issue description for a detailed explanation and analysis of the problem. This PR improves peak finding for the case where the following criteria are met:
1. The input has to be an MDEventWorkspace
2. The data source has to be of histogram-type
3. The data source has to be converted to an EventWorkspace (before it itself is then converted into an MDEventWorkspace).

**To test:**

Please find a pre-prepared sample data set here: smb://olympic/babylon5/Scratch/Anton/TEST_DATA_FIND_PEAKS_MD

## Test that UseNumberOfEventsNormalization is disabled for MDHistoworkspaces

1. Load the file `wish_event_with_peaks.nxs` and make sure the workspace has the name `wish_event_with_peaks`
2. Run the following script in order to create an MDHistoWorkspace
```python
ConvertToDiffractionMDWorkspace(InputWorkspace='wish_event_with_peaks', OutputWorkspace='wish', OneEventPerBin=False)
BinMD(InputWorkspace='wish', AlignedDim0='Q_lab_x,-11.1716,11.1716,2', AlignedDim1='Q_lab_y,-2.55985,2.55985,2', AlignedDim2='Q_lab_z,0.0104922,22.1702,2', Parallel=True, OutputWorkspace='wish_histo')
```
3. In `FindPeaksMD` set the MDHistoWorkspace as the input and check the property `UseNumberOfEventsNormalization`
4. Press `Run`
  * Confirm that an error is generated which explains that only MDEventWorkspaces can be operated with the property

## Test that NumberOfEventsNormalization provides better results for specific cases

1. Load the file `wish_event_with_peaks.nxs` and make sure the workspace has the name `wish_event_with_peaks`
2. Run the following script in order to create an MDHistoWorkspace
```python
ConvertToDiffractionMDWorkspace(InputWorkspace='wish_event_with_peaks', OutputWorkspace='wish', OneEventPerBin=False)
FindPeaksMD(InputWorkspace='wish',  MaxPeaks=100, UseNumberOfEventsNormalization=True, SignalThresholdFactor=1.2, OutputWorkspace='peaks_wish_good')
FindPeaksMD(InputWorkspace='wish', PeakDistanceThreshold=0.2, MaxPeaks=100, OutputWorkspace='peaks_wish_bad')
```
3. Plot the workspace `wish_event_with_peaks' in the InstrumentView and drag and drop the peaks workspace `peaks_wish_bad` onto it.
  * Confirm that the peaks don't match very well, ie there are many false-positives and quite a few peaks have not been found
4. Plot the workspace `wish_event_with_peaks' in the InstrumentView and drag and drop the peaks workspace `peaks_wish_good` onto it.
  * Confirm that the peaks match better. There should be considerably fewer false-positives and some of the medium-strength peaks should have been discovered.

**Release Notes** 

Please see [here](https://github.com/mantidproject/mantid/pull/19916/commits/045a7896934685c5f0895896d711d61d06bdb96a)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by NickDraper — CHANGES_REQUESTED** [review]
I think that having a combobox of peak finding (and normalization strategies) would make more sense than adding a Boolean option for this special case.

It is fairly likely that there will be more in the future.  Of course the current default should be the new default for the drop down list.

**Comment by AntonPiccardoSelg** [comment]
@NickDraper The peak finding strategy has been changed to a combobox. 
Scientific validation is complete: Pascal seems to think that the results are better than previously ( but not great for all data sets, especially if they have weak peaks and strong backgrounds)

File: docs/source/algorithms/FindPeaksMD-v1.rst [hunk 1]
```diff
@@ -45,6 +45,14 @@ is not accessible. It may give better results on
 :ref:`MDWorkspaces <MDWorkspace>`.
 
 
+For data which has originally been based on histogram-type data and that has been converted to 
+event-based data it migth be beneficial to select the NumberOfEventNormalization for the `PeakFindingStrategy` property instead of the VolumeNormalization which is the default. This
+will disable the `DensityThresholdFactor` property and enable the `SignalThresholdFactor` property.
+The algorithmic steps remain the same as above but instead of using the signal density as the sorting
+quanity the signal count (total weighted event sum divided by the number of events) is used. If 
+the events are weighted this will result in boxes with signal counts larger than one for peaks and for the majority of background counts this will 1. Hence it is possible to discrimiate between peaks and background. Note that the NumberOfEventNormalization selection of the `PeakFindingStrategy` property  does not make sense for all scenarios and infact might not produce useful results for your particalar case.
```
Comment (ThomasLohnert) [inline docs/source/algorithms/FindPeaksMD-v1.rst]: > and for the majority of background counts this will 1.

will _be_ 1?

**Review by ThomasLohnert — CHANGES_REQUESTED** [review]
The changes are a big improvement, functionality looks good. Some nitpicking in the comments

File: Framework/MDAlgorithms/src/FindPeaksMD.cpp [hunk 2]
```diff
@@ -399,7 +451,7 @@ void FindPeaksMD::findPeaks(typename MDEventWorkspace<MDE, nd>::sptr ws) {
 
       // The "bin count" used will be the box density or the number of events in
       // the box
-      double binCount = box->getSignalNormalized() * m_densityScaleFactor;
+      double binCount = box->getSignalByNEvents() * m_densityScaleFactor;
```
Comment (AntonPiccardoSelg) [inline Framework/MDAlgorithms/src/FindPeaksMD.cpp]: Good spot :+1:. Using `getSignalNormalized` makes much more sense here. At this point in the code we have already found our peaks.

**Review by ThomasLohnert — APPROVED** [review]

**Comment by AntonPiccardoSelg** [comment]
@NickDraper Your requested changes have been implemented.

**Review by NickDraper — APPROVED** [review]

File: docs/source/algorithms/FindPeaksMD-v1.rst [hunk 3]
```diff
@@ -45,6 +45,14 @@ is not accessible. It may give better results on
 :ref:`MDWorkspaces <MDWorkspace>`.
 
 
+For data which has originally been based on histogram-type data and that has been converted to 
+event-based data it migth be beneficial to select the NumberOfEventNormalization for the `PeakFindingStrategy` property instead of the VolumeNormalization which is the default. This
+will disable the `DensityThresholdFactor` property and enable the `SignalThresholdFactor` property.
+The algorithmic steps remain the same as above but instead of using the signal density as the sorting
+quanity the signal count (total weighted event sum divided by the number of events) is used. If
```
Comment (SimonHeybrock) [inline docs/source/algorithms/FindPeaksMD-v1.rst]: Typo `quanity`.

File: docs/source/algorithms/FindPeaksMD-v1.rst [hunk 4]
```diff
@@ -45,6 +45,14 @@ is not accessible. It may give better results on
 :ref:`MDWorkspaces <MDWorkspace>`.
 
 
+For data which has originally been based on histogram-type data and that has been converted to 
+event-based data it migth be beneficial to select the NumberOfEventNormalization for the `PeakFindingStrategy` property instead of the VolumeNormalization which is the default. This
+will disable the `DensityThresholdFactor` property and enable the `SignalThresholdFactor` property.
+The algorithmic steps remain the same as above but instead of using the signal density as the sorting
+quanity the signal count (total weighted event sum divided by the number of events) is used. If 
+the events are weighted this will result in boxes with signal counts larger than one for peaks and for the majority of background counts the signal count will be 1. Hence it is possible to discrimiate between peaks and background. Note that the NumberOfEventNormalization selection of the `PeakFindingStrategy` property  does not make sense for all scenarios and infact might not produce useful results for your particalar case.
```
Comment (SimonHeybrock) [inline docs/source/algorithms/FindPeaksMD-v1.rst]: Typo `discrimiate`

**Pull request merged**